In [ ]:
---
Title: 3-2 Activity
format:
    html:
        embed-resources: true

# Distances Between Observations

Read this notebook from top to bottom and fill in the code as you go. Work together and discuss with other students in the class. Try to resolve any errors on your own first, but don't get stuck; ask for help!

In addition to writing and running code, be sure to examine any output and interpret the results before moving on.

For many of these questions, there are several approaches, and there is no single right answer. You should try a few different things and compare with your classmates.

We will use `scikit-learn` extensively later, but for this activity you might want to stick with `pandas`.


In [ ]:
import pandas as pd
import numpy as np

## Ames - Recommending Similar Homes

1\. Suppose that you really like house 0 in the Ames housing data set, but it is too expensive. Find cheaper homes that are similar to it --- in terms of living area, number of bedrooms, number of bathrooms --- by calculating distances from house 0. You might want to try different distance metrics and different scaling methods; how sensitive are your results to these choices?

Be sure to actually look at the profiles of the homes that your algorithm picked out as most similar (based on these 3 variables). Do they make sense?

**_Think:_ If the goal is to find a "good deal" on a similar house, should sale price be included as a variable in your distance metric?**

In [ ]:
#Use Ames Dataset:
df_housing = pd.read_csv("https://raw.githubusercontent.com/kevindavisross/data301/main/data/AmesHousing.txt", sep="\t")
df_housing["Bathrooms"] = df_housing["Full Bath"] + df_housing["Bsmt Full Bath"] + 0.5 * (df_housing["Half Bath"] + df_housing["Bsmt Half Bath"])
df_housing_quant = df_housing[["Bedroom AbvGr", "Gr Liv Area", "Bathrooms"]]
df_housing_quant

,Bedroom AbvGr,Gr Liv Area,Bathrooms
0,3,1656,2.0
1,2,896,1.0
2,3,1329,1.5
3,3,2110,3.5
4,3,1629,2.5
...,...,...,...
2925,3,1003,2.0
2926,2,902,2.0
2927,3,970,1.5
2928,2,1389,2.0


In [ ]:
#Standardize The variables as Standard Deviations Away from the Mean
df_housing_scaled = (df_housing_quant - df_housing_quant.mean()) / df_housing_quant.std()
df_housing_scaled

,Bedroom AbvGr,Gr Liv Area,Bathrooms
0,0.176064,0.309212,-0.269988
1,-1.032058,-1.194223,-1.509056
2,0.176064,-0.337661,-0.889522
3,0.176064,1.207317,1.588613
4,0.176064,0.255801,0.349546
...,...,...,...
2925,0.176064,-0.982555,-0.269988
2926,-1.032058,-1.182354,-0.269988
2927,0.176064,-1.047836,-0.889522
2928,-1.032058,-0.218968,-0.269988


In [ ]:
# Calculate distances from house 0
dist = np.sqrt(((df_housing_scaled - df_housing_scaled.loc[0]) ** 2).sum(axis=1))
dist

,0
0,0.000000
1,2.292415
2,0.895693
3,2.064217
4,0.621832
...,...
2925,1.291768
2926,1.919460
2927,1.491778
2928,1.318534


In [ ]:
sorteddistance=dist.sort_values(ascending=True)
sorteddistance

,0
0,0.000000
1800,0.000000
731,0.003956
1128,0.007913
1226,0.009891
...,...
2180,7.340446
816,7.518242
815,7.518242
817,7.518242


In [ ]:
sorteddistance.to_frame(name='Distance').merge(df_housing[["SalePrice"]], left_index=True, right_index=True)

,Distance,SalePrice
0,0.000000,215000
1800,0.000000,249700
731,0.003956,137000
1128,0.007913,256000
1226,0.009891,165500
...,...,...
2180,7.340446,183850
816,7.518242,269500
815,7.518242,269500
817,7.518242,269500


House 731 is very similar in terms of bedrooms, bathrooms, and living area to house 0, but has a much lower price at $137,000 compared to $215,000

2\. Continuing part 1. Suppose that you really like house 0 in the data set, but it is too expensive. Find cheaper homes that are similar to it --- in terms of living area, number of bedrooms, number of bathrooms, **and House Style** --- by calculating distances from house 0. You might want to try different distance metrics and different scaling methods; how sensitive are your results to these choices?

Be sure to actually look at the profiles of the homes that your algorithm picked out as most similar. Do they make sense?

In [ ]:
# Use Ames dataset and the right variables
df_housing = pd.read_csv("https://raw.githubusercontent.com/kevindavisross/data301/main/data/AmesHousing.txt", sep="\t")
df_housing["Bathrooms"] = df_housing["Full Bath"] + df_housing["Bsmt Full Bath"] + 0.5 * (df_housing["Half Bath"] + df_housing["Bsmt Half Bath"])
df_housing_quant = df_housing[["Bedroom AbvGr", "Gr Liv Area", "Bathrooms"]]
df_housing_quant

,Bedroom AbvGr,Gr Liv Area,Bathrooms
0,3,1656,2.0
1,2,896,1.0
2,3,1329,1.5
3,3,2110,3.5
4,3,1629,2.5
...,...,...,...
2925,3,1003,2.0
2926,2,902,2.0
2927,3,970,1.5
2928,2,1389,2.0


In [ ]:
#Standardize The variables as Standard Deviations Away from the Mean
df_housing_scaled = (df_housing_quant - df_housing_quant.mean()) / df_housing_quant.std()
df_housing_scaled

,Bedroom AbvGr,Gr Liv Area,Bathrooms
0,0.176064,0.309212,-0.269988
1,-1.032058,-1.194223,-1.509056
2,0.176064,-0.337661,-0.889522
3,0.176064,1.207317,1.588613
4,0.176064,0.255801,0.349546
...,...,...,...
2925,0.176064,-0.982555,-0.269988
2926,-1.032058,-1.182354,-0.269988
2927,0.176064,-1.047836,-0.889522
2928,-1.032058,-0.218968,-0.269988


In [ ]:
#Create dummy variable for house style where 1 is 1Story and 0 is everything else
df_housing_dummies_st=pd.get_dummies(df_housing_quant, ["House Style"])*1
df_housing_dummies_st

,Bedroom AbvGr,Gr Liv Area,Bathrooms,1Story,House Style_1.5Fin,House Style_1.5Unf,House Style_1Story,House Style_2.5Fin,House Style_2.5Unf,House Style_2Story,House Style_SFoyer,House Style_SLvl
0,3,1656,2.0,1,0,0,1,0,0,0,0,0
1,2,896,1.0,1,0,0,1,0,0,0,0,0
2,3,1329,1.5,1,0,0,1,0,0,0,0,0
3,3,2110,3.5,1,0,0,1,0,0,0,0,0
4,3,1629,2.5,0,0,0,0,0,0,1,0,0
...,...,...,...,...,...,...,...,...,...,...,...,...
2925,3,1003,2.0,0,0,0,0,0,0,0,0,1
2926,2,902,2.0,1,0,0,1,0,0,0,0,0
2927,3,970,1.5,0,0,0,0,0,0,0,1,0
2928,2,1389,2.0,1,0,0,1,0,0,0,0,0


In [ ]:
final_housing=df_housing_scaled.merge(df_housing_dummies_st["House Style_1Story"].to_frame(), left_index=True, right_index=True)

In [ ]:
dist = np.sqrt(((final_housing - final_housing.loc[0]) ** 2).sum(axis=1))
dist

,0
0,0.000000
1,2.292415
2,0.895693
3,2.064217
4,1.177572
...,...
2925,1.633604
2926,1.919460
2927,1.795941
2928,1.318534


In [ ]:
sorteddistance=dist.sort_values(ascending=True)
sorteddistance

,0
0,0.000000
1800,0.000000
1128,0.007913
267,0.009891
2098,0.009891
...,...
2180,7.340446
817,7.584455
815,7.584455
816,7.584455


In [ ]:
sorteddistance.to_frame(name='Distance').merge(df_housing[["SalePrice"]], left_index=True, right_index=True)

,Distance,SalePrice
0,0.000000,215000
1800,0.000000,249700
1128,0.007913,256000
267,0.009891,224900
2098,0.009891,250000
...,...,...
2180,7.340446,183850
817,7.584455,269500
815,7.584455,269500
816,7.584455,269500


House 1940 is very similar in terms of bedrooms, bathrooms, and living area to house 0 (and has the same style), but has a much lower price at $153,000 compared to $215,000

3\. Continuing parts 1 and 2. Suppose that you really like house 0 in the data set, but it is too expensive. Find cheaper homes that are similar to it, by calculating distances. You can **choose the variables to include, but include both quantitative and categorical variables**. Be sure to actually look at the profiles of the homes that your algorithm picked out as most similar. Do they make sense?

You might want to try different distance metrics and different scaling methods; how sensitive are your results to these choices?

_Hint:_ There are many variables in the data set. Do not attempt to compute distance based on all the variables! You will want to pare down the number of variables, but be sure to include a mixture of categorical and quantitative variables. Refer to the [data documentation](https://ww2.amstat.org/publications/jse/v19n3/decock/DataDocumentation.txt) for information about the variables.


In [ ]:
# Import Dataset and chosen variables
df_housing = pd.read_csv("https://raw.githubusercontent.com/kevindavisross/data301/main/data/AmesHousing.txt", sep="\t")
df_housing["Bathrooms"] = df_housing["Full Bath"] + df_housing["Bsmt Full Bath"] + 0.5 * (df_housing["Half Bath"] + df_housing["Bsmt Half Bath"])
df_housing_quant = df_housing[["Bedroom AbvGr", "Overall Cond", "Bathrooms"]]
df_housing_quant

,Bedroom AbvGr,Overall Cond,Bathrooms
0,3,5,2.0
1,2,6,1.0
2,3,6,1.5
3,3,5,3.5
4,3,5,2.5
...,...,...,...
2925,3,6,2.0
2926,2,5,2.0
2927,3,5,1.5
2928,2,5,2.0


In [ ]:
#Standardize The variables as Standard Deviations Away from the Mean
df_housing_scaled = (df_housing_quant - df_housing_quant.mean()) / df_housing_quant.std()
df_housing_scaled

,Bedroom AbvGr,Overall Cond,Bathrooms
0,0.176064,-0.506632,-0.269988
1,-1.032058,0.393024,-1.509056
2,0.176064,0.393024,-0.889522
3,0.176064,-0.506632,1.588613
4,0.176064,-0.506632,0.349546
...,...,...,...
2925,0.176064,0.393024,-0.269988
2926,-1.032058,-0.506632,-0.269988
2927,0.176064,-0.506632,-0.889522
2928,-1.032058,-0.506632,-0.269988


In [ ]:
df_housing_dummies_st=pd.get_dummies(df_housing, columns=["Neighborhood"])*1
df_housing_dummies_st

,Order,PID,MS SubClass,MS Zoning,Lot Frontage,Lot Area,Street,Alley,Lot Shape,Land Contour,...,Neighborhood_NoRidge,Neighborhood_NridgHt,Neighborhood_OldTown,Neighborhood_SWISU,Neighborhood_Sawyer,Neighborhood_SawyerW,Neighborhood_Somerst,Neighborhood_StoneBr,Neighborhood_Timber,Neighborhood_Veenker
0,1,526301100,20,RL,141.0,31770,Pave,NaN,IR1,Lvl,...,0,0,0,0,0,0,0,0,0,0
1,2,526350040,20,RH,80.0,11622,Pave,NaN,Reg,Lvl,...,0,0,0,0,0,0,0,0,0,0
2,3,526351010,20,RL,81.0,14267,Pave,NaN,IR1,Lvl,...,0,0,0,0,0,0,0,0,0,0
3,4,526353030,20,RL,93.0,11160,Pave,NaN,Reg,Lvl,...,0,0,0,0,0,0,0,0,0,0
4,5,527105010,60,RL,74.0,13830,Pave,NaN,IR1,Lvl,...,0,0,0,0,0,0,0,0,0,0
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
2925,2926,923275080,80,RL,37.0,7937,Pave,NaN,IR1,Lvl,...,0,0,0,0,0,0,0,0,0,0
2926,2927,923276100,20,RL,NaN,8885,Pave,NaN,IR1,Low,...,0,0,0,0,0,0,0,0,0,0
2927,2928,923400125,85,RL,62.0,10441,Pave,NaN,Reg,Lvl,...,0,0,0,0,0,0,0,0,0,0
2928,2929,924100070,20,RL,77.0,10010,Pave,NaN,Reg,Lvl,...,0,0,0,0,0,0,0,0,0,0


In [ ]:
final_housing=df_housing_scaled.merge(df_housing_dummies_st["Neighborhood_NAmes"].to_frame(), left_index=True, right_index=True)

In [ ]:
dist = np.sqrt(((final_housing - final_housing.loc[0]) ** 2).sum(axis=1))
dist

,0
0,0.000000
1,1.950443
2,1.092338
3,1.858601
4,1.176360
...,...
2925,1.345132
2926,1.568298
2927,1.176360
2928,1.568298


In [ ]:
sorteddistance=dist.sort_values(ascending=True)
sorteddistance

,0
0,0.000000
2607,0.000000
2606,0.000000
2375,0.000000
1244,0.000000
...,...
2723,5.241176
2194,6.381730
815,7.246946
817,7.246946


In [ ]:
sorteddistance.to_frame(name='Distance').merge(df_housing[["SalePrice"]], left_index=True, right_index=True)

,Distance,SalePrice
0,0.000000,215000
2607,0.000000,159950
2606,0.000000,160000
2375,0.000000,123000
1244,0.000000,133000
...,...,...
2723,5.241176,145000
2194,6.381730,200000
815,7.246946,269500
817,7.246946,269500


I would reccomend house 2375.  It is in the same neighborhood as house 0, has the same number of bedrooms and bathrooms, and is in the same condition, but only costs $123,000 in comparison to the house 0 cost of $215,000

## Colleges similar to Cal Poly

We'll use data from the [College Scorecard data](https://collegescorecard.ed.gov/) to find colleges and universities that are similar to Cal Poly.

In [ ]:
df_college = pd.read_csv("https://datasci112.stanford.edu/data/college_attributes.csv")

df_college.set_index("Institution", inplace = True)

df_college

,City,State,AdmissionRate,Undergraduates,CarnegieClassification,Ownership,PCIP01,PCIP03,PCIP04,PCIP05,...,PCIP44,PCIP45,PCIP46,PCIP47,PCIP48,PCIP49,PCIP50,PCIP51,PCIP52,PCIP54
Institution,,,,,,,,,,,,,,,,,,,,,
Alabama A & M University,Normal,AL,0.7160,5098.0,Master's Colleges & Universities: Larger Programs,Public,0.0445,0.0071,0.0053,0.0000,...,0.0409,0.0249,0.0,0.0,0.0,0.0,0.0231,0.0000,0.1637,0.0000
University of Alabama at Birmingham,Birmingham,AL,0.8854,13284.0,Doctoral Universities: Very High Research Acti...,Public,0.0000,0.0000,0.0000,0.0020,...,0.0195,0.0239,0.0,0.0,0.0,0.0,0.0249,0.2088,0.2159,0.0141
University of Alabama in Huntsville,Huntsville,AL,0.7367,7358.0,Doctoral Universities: Very High Research Acti...,Public,0.0000,0.0000,0.0000,0.0000,...,0.0000,0.0127,0.0,0.0,0.0,0.0,0.0407,0.1341,0.1930,0.0073
Alabama State University,Montgomery,AL,0.9799,3495.0,Doctoral/Professional Universities,Public,0.0000,0.0000,0.0000,0.0000,...,0.0648,0.0196,0.0,0.0,0.0,0.0,0.0511,0.0904,0.1513,0.0059
The University of Alabama,Tuscaloosa,AL,0.7890,30725.0,Doctoral Universities: Very High Research Acti...,Public,0.0000,0.0061,0.0000,0.0019,...,0.0072,0.0661,0.0,0.0,0.0,0.0,0.0234,0.1077,0.2916,0.0096
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
Florida Academy of Nursing,Miramar,FL,0.3088,239.0,Not applicable,Private for-profit,0.0000,0.0000,0.0000,0.0000,...,0.0000,0.0000,0.0,0.0,0.0,0.0,0.0000,1.0000,0.0000,0.0000
Herzing University-Tampa,Tampa,FL,0.9630,68.0,Not applicable,Private nonprofit,0.0000,0.0000,0.0000,0.0000,...,0.0000,0.0000,0.0,0.0,0.0,0.0,0.0000,0.0000,0.0000,0.0000
Abilene Christian University-Undergraduate Online,Addison,TX,1.0000,415.0,Not applicable,Private nonprofit,0.0000,0.0000,0.0000,0.0000,...,0.0000,0.0000,0.0,0.0,0.0,0.0,0.0000,0.0000,0.0000,0.0000


We'll want to single out Cal Poly, which we can do like this.

In [ ]:
school_name = "California Polytechnic State University-San Luis Obispo"

cp = df_college.loc[school_name]

cp

,California Polytechnic State University-San Luis Obispo
City,San Luis Obispo
State,CA
AdmissionRate,0.33
Undergraduates,21090.0
CarnegieClassification,Master's Colleges & Universities: Larger Programs
Ownership,Public
PCIP01,0.1084
PCIP03,0.0255
PCIP04,0.0441
PCIP05,0.0019


1\. Based on only the admission rate and the number of undergraduates, what schools are most similar to Cal Poly? Specify how you're making this decision.

In [ ]:
df_college_quant=df_college[["AdmissionRate", "Undergraduates"]]
df_college_quant

,AdmissionRate,Undergraduates
Institution,,
Alabama A & M University,0.7160,5098.0
University of Alabama at Birmingham,0.8854,13284.0
University of Alabama in Huntsville,0.7367,7358.0
Alabama State University,0.9799,3495.0
The University of Alabama,0.7890,30725.0
...,...,...
Florida Academy of Nursing,0.3088,239.0
Herzing University-Tampa,0.9630,68.0
Abilene Christian University-Undergraduate Online,1.0000,415.0


In [ ]:
#Standardize the variables
df_college_scaled = (df_college_quant - df_college_quant.mean()) / df_college_quant.std()
df_college_scaled

,AdmissionRate,Undergraduates
Institution,,
Alabama A & M University,-0.071116,0.121332
University of Alabama at Birmingham,0.695587,1.175104
University of Alabama in Huntsville,0.022572,0.412258
Alabama State University,1.123294,-0.085020
The University of Alabama,0.259282,3.420258
...,...,...
Florida Academy of Nursing,-1.914100,-0.504160
Herzing University-Tampa,1.046804,-0.526173
Abilene Christian University-Undergraduate Online,1.214266,-0.481504


In [ ]:
dist=np.sqrt((df_college_scaled-df_college_scaled.loc[school_name]) ** 2).sum(axis=1)
dist

,0
Institution,
Alabama A & M University,3.805660
University of Alabama at Birmingham,3.518591
University of Alabama in Huntsville,3.608422
Alabama State University,5.206421
The University of Alabama,3.317730
...,...
Florida Academy of Nursing,2.780070
Herzing University-Tampa,5.571085
Abilene Christian University-Undergraduate Online,5.693878


In [ ]:
sorteddistance=dist.sort_values(ascending=True)
sorteddistance

,0
Institution,
California Polytechnic State University-San Luis Obispo,0.000000
University of California-Santa Barbara,0.429191
DeVry University-Illinois,0.741854
University of North Carolina at Chapel Hill,0.746376
Clemson University,0.796807
...,...
Liberty University,6.358227
Grand Canyon University,7.806530
Arizona State University Campus Immersion,8.049443


UCSB is the most similar to Cal Poly in terms of undergraduates and Admission Rate

2\. Now consider the admission rate, the number of undergraduates, and also the [Carnegie classification](https://en.wikipedia.org/wiki/Carnegie_Classification_of_Institutions_of_Higher_Education) of the type of school, and the ownership (public, private, etc.) Based on these variables, what schools are most similar to Cal Poly? Specify how you're making this decision.

In [ ]:
df_college_quant=df_college[["AdmissionRate", "Undergraduates"]]
df_college_quant

,AdmissionRate,Undergraduates
Institution,,
Alabama A & M University,0.7160,5098.0
University of Alabama at Birmingham,0.8854,13284.0
University of Alabama in Huntsville,0.7367,7358.0
Alabama State University,0.9799,3495.0
The University of Alabama,0.7890,30725.0
...,...,...
Florida Academy of Nursing,0.3088,239.0
Herzing University-Tampa,0.9630,68.0
Abilene Christian University-Undergraduate Online,1.0000,415.0


In [ ]:
#Standardize the variables
df_college_scaled = (df_college_quant - df_college_quant.mean()) / df_college_quant.std()
df_college_scaled

,AdmissionRate,Undergraduates
Institution,,
Alabama A & M University,-0.071116,0.121332
University of Alabama at Birmingham,0.695587,1.175104
University of Alabama in Huntsville,0.022572,0.412258
Alabama State University,1.123294,-0.085020
The University of Alabama,0.259282,3.420258
...,...,...
Florida Academy of Nursing,-1.914100,-0.504160
Herzing University-Tampa,1.046804,-0.526173
Abilene Christian University-Undergraduate Online,1.214266,-0.481504


In [ ]:
df_college_dummies_ownership=pd.get_dummies(df_college, columns=["Ownership"])*1
df_college_dummies_ownership

,City,State,AdmissionRate,Undergraduates,CarnegieClassification,PCIP01,PCIP03,PCIP04,PCIP05,PCIP09,...,PCIP47,PCIP48,PCIP49,PCIP50,PCIP51,PCIP52,PCIP54,Ownership_Private for-profit,Ownership_Private nonprofit,Ownership_Public
Institution,,,,,,,,,,,,,,,,,,,,,
Alabama A & M University,Normal,AL,0.7160,5098.0,Master's Colleges & Universities: Larger Programs,0.0445,0.0071,0.0053,0.0000,0.0000,...,0.0,0.0,0.0,0.0231,0.0000,0.1637,0.0000,0,0,1
University of Alabama at Birmingham,Birmingham,AL,0.8854,13284.0,Doctoral Universities: Very High Research Acti...,0.0000,0.0000,0.0000,0.0020,0.0333,...,0.0,0.0,0.0,0.0249,0.2088,0.2159,0.0141,0,0,1
University of Alabama in Huntsville,Huntsville,AL,0.7367,7358.0,Doctoral Universities: Very High Research Acti...,0.0000,0.0000,0.0000,0.0000,0.0140,...,0.0,0.0,0.0,0.0407,0.1341,0.1930,0.0073,0,0,1
Alabama State University,Montgomery,AL,0.9799,3495.0,Doctoral/Professional Universities,0.0000,0.0000,0.0000,0.0000,0.0923,...,0.0,0.0,0.0,0.0511,0.0904,0.1513,0.0059,0,0,1
The University of Alabama,Tuscaloosa,AL,0.7890,30725.0,Doctoral Universities: Very High Research Acti...,0.0000,0.0061,0.0000,0.0019,0.0877,...,0.0,0.0,0.0,0.0234,0.1077,0.2916,0.0096,0,0,1
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
Florida Academy of Nursing,Miramar,FL,0.3088,239.0,Not applicable,0.0000,0.0000,0.0000,0.0000,0.0000,...,0.0,0.0,0.0,0.0000,1.0000,0.0000,0.0000,1,0,0
Herzing University-Tampa,Tampa,FL,0.9630,68.0,Not applicable,0.0000,0.0000,0.0000,0.0000,0.0000,...,0.0,0.0,0.0,0.0000,0.0000,0.0000,0.0000,0,1,0
Abilene Christian University-Undergraduate Online,Addison,TX,1.0000,415.0,Not applicable,0.0000,0.0000,0.0000,0.0000,0.0000,...,0.0,0.0,0.0,0.0000,0.0000,0.0000,0.0000,0,1,0


In [ ]:
df_college_dummies_type=pd.get_dummies(df_college, columns=["CarnegieClassification"])*1
df_college_dummies_type

,City,State,AdmissionRate,Undergraduates,Ownership,PCIP01,PCIP03,PCIP04,PCIP05,PCIP09,...,CarnegieClassification_Special Focus Four-Year: Faith-Related Institutions,CarnegieClassification_Special Focus Four-Year: Medical Schools & Centers,CarnegieClassification_Special Focus Four-Year: Other Health Professions Schools,CarnegieClassification_Special Focus Four-Year: Other Special Focus Institutions,CarnegieClassification_Special Focus Four-Year: Research Institution,CarnegieClassification_Special Focus Two-Year: Arts & Design,CarnegieClassification_Special Focus Two-Year: Health Professions,CarnegieClassification_Special Focus Two-Year: Other Fields,CarnegieClassification_Special Focus Two-Year: Technical Professions,CarnegieClassification_Tribal Colleges
Institution,,,,,,,,,,,,,,,,,,,,,
Alabama A & M University,Normal,AL,0.7160,5098.0,Public,0.0445,0.0071,0.0053,0.0000,0.0000,...,0,0,0,0,0,0,0,0,0,0
University of Alabama at Birmingham,Birmingham,AL,0.8854,13284.0,Public,0.0000,0.0000,0.0000,0.0020,0.0333,...,0,0,0,0,0,0,0,0,0,0
University of Alabama in Huntsville,Huntsville,AL,0.7367,7358.0,Public,0.0000,0.0000,0.0000,0.0000,0.0140,...,0,0,0,0,0,0,0,0,0,0
Alabama State University,Montgomery,AL,0.9799,3495.0,Public,0.0000,0.0000,0.0000,0.0000,0.0923,...,0,0,0,0,0,0,0,0,0,0
The University of Alabama,Tuscaloosa,AL,0.7890,30725.0,Public,0.0000,0.0061,0.0000,0.0019,0.0877,...,0,0,0,0,0,0,0,0,0,0
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
Florida Academy of Nursing,Miramar,FL,0.3088,239.0,Private for-profit,0.0000,0.0000,0.0000,0.0000,0.0000,...,0,0,0,0,0,0,0,0,0,0
Herzing University-Tampa,Tampa,FL,0.9630,68.0,Private nonprofit,0.0000,0.0000,0.0000,0.0000,0.0000,...,0,0,0,0,0,0,0,0,0,0
Abilene Christian University-Undergraduate Online,Addison,TX,1.0000,415.0,Private nonprofit,0.0000,0.0000,0.0000,0.0000,0.0000,...,0,0,0,0,0,0,0,0,0,0


In [63]:
inter_college=df_college_scaled.merge(df_college_dummies_ownership["Ownership_Public"].to_frame(), left_index=True, right_index=True)

In [64]:
inter_college

,AdmissionRate,Undergraduates,Ownership_Public
Institution,,,
Alabama A & M University,-0.071116,0.121332,1
University of Alabama at Birmingham,0.695587,1.175104,1
University of Alabama in Huntsville,0.022572,0.412258,1
Alabama State University,1.123294,-0.085020,1
The University of Alabama,0.259282,3.420258,1
...,...,...,...
Florida Academy of Nursing,-1.914100,-0.504160,0
Herzing University-Tampa,1.046804,-0.526173,0
Abilene Christian University-Undergraduate Online,1.214266,-0.481504,0


In [66]:
final_college=inter_college.merge(df_college_dummies_type["CarnegieClassification_Master's Colleges & Universities: Larger Programs"].to_frame(), left_index=True, right_index=True)

In [67]:
final_college

,AdmissionRate,Undergraduates,Ownership_Public,CarnegieClassification_Master's Colleges & Universities: Larger Programs
Institution,,,,
Alabama A & M University,-0.071116,0.121332,1,1
University of Alabama at Birmingham,0.695587,1.175104,1,0
University of Alabama in Huntsville,0.022572,0.412258,1,0
Alabama State University,1.123294,-0.085020,1,0
The University of Alabama,0.259282,3.420258,1,0
...,...,...,...,...
Florida Academy of Nursing,-1.914100,-0.504160,0,0
Herzing University-Tampa,1.046804,-0.526173,0,0
Abilene Christian University-Undergraduate Online,1.214266,-0.481504,0,0


In [68]:
dist=np.sqrt((final_college-final_college.loc[school_name]) ** 2).sum(axis=1)
dist

,0
Institution,
Alabama A & M University,3.805660
University of Alabama at Birmingham,4.518591
University of Alabama in Huntsville,4.608422
Alabama State University,6.206421
The University of Alabama,4.317730
...,...
Florida Academy of Nursing,4.780070
Herzing University-Tampa,7.571085
Abilene Christian University-Undergraduate Online,7.693878


In [69]:
sorteddistance=dist.sort_values(ascending=True)
sorteddistance

,0
Institution,
California Polytechnic State University-San Luis Obispo,0.000000
CUNY Hunter College,1.072635
University of California-Santa Barbara,1.429191
CUNY Bernard M Baruch College,1.516545
CUNY John Jay College of Criminal Justice,1.586893
...,...
Liberty University,8.358227
Arizona State University Campus Immersion,9.049443
Grand Canyon University,9.806530


CUNY Hunter College, UCSB, and CUNY Bernard M Barush College are the three most similar universities to Cal Poly by these metrics.

3\. The columns whose names begin with "PCIP" contain the proportions of students at each school studying various fields (e.g., Engineering, Psychology). Each field is represented by a two-digit code called the [CIP code](https://nces.ed.gov/ipeds/cipcode/browse.aspx?y=55).

If we only consider the proportions of students studying various fields, what schools are most similar to Cal Poly? Specify how you're making this decision.

In [70]:
# Create datafram with institution and all PCIP's
df_college_PCIP=df_college[df_college.columns[df_college.columns.str.startswith("PCIP")]]
df_college_PCIP

,PCIP01,PCIP03,PCIP04,PCIP05,PCIP09,PCIP10,PCIP11,PCIP12,PCIP13,PCIP14,...,PCIP44,PCIP45,PCIP46,PCIP47,PCIP48,PCIP49,PCIP50,PCIP51,PCIP52,PCIP54
Institution,,,,,,,,,,,,,,,,,,,,,
Alabama A & M University,0.0445,0.0071,0.0053,0.0000,0.0000,0.0285,0.0658,0.0,0.0391,0.1210,...,0.0409,0.0249,0.0,0.0,0.0,0.0,0.0231,0.0000,0.1637,0.0000
University of Alabama at Birmingham,0.0000,0.0000,0.0000,0.0020,0.0333,0.0000,0.0229,0.0,0.0667,0.0559,...,0.0195,0.0239,0.0,0.0,0.0,0.0,0.0249,0.2088,0.2159,0.0141
University of Alabama in Huntsville,0.0000,0.0000,0.0000,0.0000,0.0140,0.0000,0.0692,0.0,0.0218,0.3028,...,0.0000,0.0127,0.0,0.0,0.0,0.0,0.0407,0.1341,0.1930,0.0073
Alabama State University,0.0000,0.0000,0.0000,0.0000,0.0923,0.0000,0.0530,0.0,0.0589,0.0138,...,0.0648,0.0196,0.0,0.0,0.0,0.0,0.0511,0.0904,0.1513,0.0059
The University of Alabama,0.0000,0.0061,0.0000,0.0019,0.0877,0.0000,0.0148,0.0,0.0262,0.1198,...,0.0072,0.0661,0.0,0.0,0.0,0.0,0.0234,0.1077,0.2916,0.0096
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
Florida Academy of Nursing,0.0000,0.0000,0.0000,0.0000,0.0000,0.0000,0.0000,0.0,0.0000,0.0000,...,0.0000,0.0000,0.0,0.0,0.0,0.0,0.0000,1.0000,0.0000,0.0000
Herzing University-Tampa,0.0000,0.0000,0.0000,0.0000,0.0000,0.0000,0.0000,0.0,0.0000,0.0000,...,0.0000,0.0000,0.0,0.0,0.0,0.0,0.0000,0.0000,0.0000,0.0000
Abilene Christian University-Undergraduate Online,0.0000,0.0000,0.0000,0.0000,0.0000,0.0000,0.0000,0.0,0.0000,0.0000,...,0.0000,0.0000,0.0,0.0,0.0,0.0,0.0000,0.0000,0.0000,0.0000


In [72]:
#Standardize the variables
df_college_PCIP_scaled = (df_college_PCIP - df_college_PCIP.mean()) / df_college_PCIP.std()
df_college_PCIP

,PCIP01,PCIP03,PCIP04,PCIP05,PCIP09,PCIP10,PCIP11,PCIP12,PCIP13,PCIP14,...,PCIP44,PCIP45,PCIP46,PCIP47,PCIP48,PCIP49,PCIP50,PCIP51,PCIP52,PCIP54
Institution,,,,,,,,,,,,,,,,,,,,,
Alabama A & M University,0.597422,-0.040470,0.038251,-0.129573,-0.706682,0.77606,0.407050,-0.113381,-0.075358,1.055962,...,0.837379,-0.295171,-0.078362,-0.100123,-0.067964,-0.118704,-0.236661,-0.671477,0.149548,-0.711122
University of Alabama at Birmingham,-0.166825,-0.283732,-0.102031,-0.047792,0.159480,-0.13387,-0.150064,-0.113381,0.253826,0.302637,...,0.170217,-0.310606,-0.078362,-0.100123,-0.067964,-0.118704,-0.225923,0.023221,0.524549,0.306615
University of Alabama in Huntsville,-0.166825,-0.283732,-0.102031,-0.129573,-0.342530,-0.13387,0.451203,-0.113381,-0.281694,3.159717,...,-0.437712,-0.483471,-0.078362,-0.100123,-0.067964,-0.118704,-0.131664,-0.225313,0.360037,-0.184208
Alabama State University,-0.166825,-0.283732,-0.102031,-0.129573,1.694120,-0.13387,0.240825,-0.113381,0.160796,-0.184536,...,1.582482,-0.376974,-0.078362,-0.100123,-0.067964,-0.118704,-0.069620,-0.370707,0.060467,-0.285260
The University of Alabama,-0.166825,-0.074733,-0.102031,-0.051881,1.574470,-0.13387,-0.255254,-0.113381,-0.229216,1.042076,...,-0.213246,0.340727,-0.078362,-0.100123,-0.067964,-0.118704,-0.234871,-0.313149,1.068372,-0.018194
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
Florida Academy of Nursing,-0.166825,-0.283732,-0.102031,-0.129573,-0.706682,-0.13387,-0.447451,-0.113381,-0.541702,-0.344227,...,-0.437712,-0.679489,-0.078362,-0.100123,-0.067964,-0.118704,-0.374470,2.655623,-1.026461,-0.711122
Herzing University-Tampa,-0.166825,-0.283732,-0.102031,-0.129573,-0.706682,-0.13387,-0.447451,-0.113381,-0.541702,-0.344227,...,-0.437712,-0.679489,-0.078362,-0.100123,-0.067964,-0.118704,-0.374470,-0.671477,-1.026461,-0.711122
Abilene Christian University-Undergraduate Online,-0.166825,-0.283732,-0.102031,-0.129573,-0.706682,-0.13387,-0.447451,-0.113381,-0.541702,-0.344227,...,-0.437712,-0.679489,-0.078362,-0.100123,-0.067964,-0.118704,-0.374470,-0.671477,-1.026461,-0.711122


In [73]:
dist=np.sqrt((df_college_PCIP_scaled-df_college_PCIP_scaled.loc[school_name]) ** 2).sum(axis=1)
dist

,0
Institution,
Alabama A & M University,17.454328
University of Alabama at Birmingham,15.688193
University of Alabama in Huntsville,12.013726
Alabama State University,20.202156
The University of Alabama,14.607759
...,...
Florida Academy of Nursing,21.381016
Herzing University-Tampa,18.093841
Abilene Christian University-Undergraduate Online,18.093841


In [74]:
sorteddistance=dist.sort_values(ascending=True)
sorteddistance

,0
Institution,
California Polytechnic State University-San Luis Obispo,0.000000
Iowa State University,7.559127
California State Polytechnic University-Pomona,7.710675
Texas A & M University-College Station,8.063105
North Carolina State University at Raleigh,8.501392
...,...
Orleans Technical College,53.718774
Patrick Henry College,53.724344
University of Southern Mississippi,55.685468


Iowa state university and Cal Poly Pamona are the most similar to Cal Poly when it comes to the proportion of students across all the various programs in the dataset